chirp/drivers/tdh8.py

```diff
@@ -0,0 +1,2193 @@
+        dtmf.append(rs)
+
+        gcode_list_4 = self._memobj.group4.group4
+        used_group4 = ''
+        for i in gcode_list_4:
+            if i == 0xFF:
+                continue
+            used_group4 += str(i)[3]
+        group4_val = RadioSettingValueString(0, 7, used_group4)
+        rs = RadioSetting("group4", "4", group4_val)
+        group4_val.set_charset(dtmfcharsani)
+        dtmf.append(rs)
+
+        gcode_list_5 = self._memobj.group5.group5
+        used_group5 = ''
+        for i in gcode_list_5:
+            if i == 0xFF:
+                continue
+            used_group5 += str(i)[3]
+        group5_val = RadioSettingValueString(0, 7, used_group5)
+        rs = RadioSetting("group5", "5", group5_val)
+        group5_val.set_charset(dtmfcharsani)
+        dtmf.append(rs)
+
+        gcode_list_6 = self._memobj.group6.group6
+        used_group6 = ''
+        for i in gcode_list_6:
+            if i == 0xFF:
+                continue
+            used_group6 += str(i)[3]
+        group6_val = RadioSettingValueString(0, 7, used_group6)
+        rs = RadioSetting("group6", "6", group6_val)
+        group6_val.set_charset(dtmfcharsani)
+        dtmf.append(rs)
+
+        gcode_list_7 = self._memobj.group7.group7
+        used_group7 = ''
+        for i in gcode_list_7:
+            if i == 0xFF:
+                continue
+            used_group7 += str(i)[3]
+        group7_val = RadioSettingValueString(0, 7, used_group7)
+        rs = RadioSetting("group7", "7", group7_val)
+        group7_val.set_charset(dtmfcharsani)
+        dtmf.append(rs)
+
+        gcode_list_8 = self._memobj.group8.group8
+        used_group8 = ''
+        for i in gcode_list_8:
+            if i == 0xFF:
+                continue
+            used_group8 += str(i)[3]
+        group8_val = RadioSettingValueString(0, 7, used_group8)
+        rs = RadioSetting("group8", "8", group7_val)
+        group8_val.set_charset(dtmfcharsani)
+        dtmf.append(rs)
+
+        scode_list = self._memobj.startcode.scode
+        used_scode = ''
+        for i in scode_list:
+            if i == 0xFF:
+                continue
+            used_scode += str(i)[3]
+        scode_val = RadioSettingValueString(0, 7, used_scode)
+        rs = RadioSetting("scode", "PTT ID Starting(BOT)", scode_val)
+        scode_val.set_charset(dtmfcharsani)
+        dtmf.append(rs)
+
+        ecode_list = self._memobj.endcode.ecode
+        used_ecode = ''
+        for i in ecode_list:
+            if i == 0xFF:
+                continue
+            used_ecode += str(i)[3]
+        ecode_val = RadioSettingValueString(0, 7, used_ecode)
+        rs = RadioSetting("ecode", "PTT ID Ending(BOT)", ecode_val)
+        dtmf.append(rs)
+
+        return group
+
+    def get_settings(self):
+        try:
+            return self._get_settings()
+        except:
+            import traceback
+            LOG.error("Failed to parse settings: %s", traceback.format_exc())
```

Comment:
I think I mentioned this before, but please do not import and format your own traceback use `LOG.exception` which will include the stack trace automatically.